scripts/gen_dockerfile_test.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/env python3
+
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit test for gen_dockerfile.py"""
+
+import argparse
+import os
+import re
+import shutil
+import subprocess
+import tempfile
+import unittest
+import unittest.mock
+
+import yaml
+
+import gen_dockerfile
+
+# Expected list of files generated
+EXPECTED_OUTPUT_FILES = ['Dockerfile', '.dockerignore']
+
+
+class GenDockerfileTest(unittest.TestCase):
```

Comment:
> I'm sorry that you view it so negatively, as such a drag and failure.

It's not that, it's just that it's designed to deal with conditions that don't match up to the external world. I think stuff in google3-style works great for Google. I do not think it works great for the outside world just because of the impedance mismatch.

> Replace "Goobuntu" with "Ubuntu (Desktop Edition)".

That's maybe a fine constraint if our target users were sysadmin that are solely limited to what's available (and vetted) by an OS packager.

---

Basically, there's three audiences we need to consider here:

1. Our users - the people who directly consume this software. Our users actually consume this particular code through a docker container that we create that runs on infrastructure we control. They consume it solely through that interface (afaik) - so it seems like we can really do whatever we want here - our software is two levels removed from the user - we're essentially writing a plugin to a build system.
2. The maintainers - this is us. This is, from my perspective, the current audience you're considering the most here, which is fine in this case because the end users don't directly interact with this cod. You're choosing tools and software that are most comfortable and familiar to yourself. You're optimizing for you being the sole maintainer.
3. Contributors. This is me and random customers/users who are brave enough to want to learn how things work and give back. You may or may not consider this audience important. However, if you do consider this audience important you must carefully weigh this audience's expectations against your biases in (2). If they are at odds, you gotta figure out if you're more comfortable with sacrificing some of your comfort in (2) in exchange for some comfort here. You may decide that it's not worth it and that you feel like your way is the one true way and that going against the grain is worth the potential cost of a few contributions for this project.

It is totally up to you how you want to do this. Some projects choose to be open source but close off all external contributions. That's a valid strategy. Some projects bend over backwards to make themselves more aligned with the community so that external contributions are easy. There's no universal answer for every project.

All this is to say I just want you to be aware that some things have changed a bit in the Python community that can make things easier for folks. And now that you're aware, it's your decision on whether or not you want this project to use those things.